pyro/ops/newton.py

```diff
@@ -0,0 +1,90 @@
+from __future__ import absolute_import, division, print_function
+
+import warnings
+
+import torch
+from torch.autograd import grad
+
+
+def _warn_if_nan(tensor, name):
+    if torch.isnan(tensor).any():
+        warnings.warn('Encountered nan elements in {}'.format(name))
+
+
+def newton_step_2d(loss, x, trust_radius=None):
+    """
+    Performs a Newton update step to minimize loss on a batch of 2-dimensional
+    variables, optionally regularizing to constrain to a trust region.
+
+    ``loss`` must be twice-differentiable as a function of ``x``. If ``loss``
+    is ``2+d``-times differentiable, then the return value of this function is
+    ``d``-times differentiable.
+
+    When ``loss`` is interpreted as a negative log probability density, then
+    the return value of this function can be used to construct a Laplace
+    approximation ``MultivariateNormal(mode,cov)``.
+
+    .. warning:: Take care to detach the result of this function when used in
+        an optimization loop. If you forget to detach the result of this
```

Comment:
Is there any extra protection we could/should add to guard against this?

Reply:
The only way I see to add protection is to pass in a `loss_fn` callback and perform those ops for the user, but that feels quite inflexible. I prefer the unix style simple-function-with-provisos.